pkg/docker/image.go

```diff
@@ -0,0 +1,87 @@
+package docker
+
+// Create docker images with a customized docker context.
+
+import (
+	"context"
+	"os"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/archive"
+
+	"go.uber.org/zap"
+)
+
+type EnsureImageOpts struct {
+	Name      string                   // reuired for EnsureImage
+	BuildCtx  string                   // required
+	BuildOpts *types.ImageBuildOptions // optional
+}
+
+func defaultBuildOptsFor(name string) *types.ImageBuildOptions {
+	return &types.ImageBuildOptions{
+		SuppressOutput: false,
+		Remove:         true,
+		ForceRemove:    true,
+		PullParent:     true,
+		Dockerfile:     "/Dockerfile",
+		Tags:           []string{name},
+	}
+}
+
+// buildImage
+// Use a Dockerfile and supporting files to build a docker image.
+// Think `docker build /path/to/build`
+func BuildImage(ctx context.Context, client *client.Client, opts *EnsureImageOpts) error {
+	buildCtx, err := archive.TarWithOptions(opts.BuildCtx, &archive.TarOptions{})
+	if err != nil {
+		return err
+	}
+	defer buildCtx.Close()
+
+	var buildOpts *types.ImageBuildOptions
+	if opts.BuildOpts == nil {
+		buildOpts = defaultBuildOptsFor(opts.Name)
+	} else {
+		buildOpts = opts.BuildOpts
+	}
+
+	buildResponse, err := client.ImageBuild(ctx, buildCtx, *buildOpts)
+	if err != nil {
+		return err
+	}
+	defer buildResponse.Body.Close()
+	return PipeOutput(buildResponse.Body, os.Stdout)
+}
+
+// EnsureImage
+// Create a docker image from build context.
+// If an image with the requested tag already exists, don't re-create it.
```

Comment:
These godocs seem like notes you wrote for yourself rather than documentation for API users? I'd try with:

> EnsureImage checks if the specified image exists. If it does, it returns immediately with `created=false`. Otherwise, it builds the image using the supplied `BuildImageOpts` parameters, and returns `created=true`, along with any errors that may have occurred during the process.